File: app/src/main/java/top/itning/yunshuclassschedule/ui/fragment/ClassScheduleFragment.java
package top.itning.yunshuclassschedule.ui.fragment;

import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.preference.PreferenceManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import top.itning.yunshuclassschedule.R;
import top.itning.yunshuclassschedule.entity.EventEntity;
import top.itning.yunshuclassschedule.util.ThemeChangeUtil;

import static top.itning.yunshuclassschedule.ui.fragment.setting.SettingsFragment.DEFAULT_SHOW_MAIN_FRAGMENT;

/**
 * 课程表
 *
 * @author itning
 */
public class ClassScheduleFragment extends Fragment {
    private static final String TAG = "ClassScheduleFragment";

    private static final String TODAY = "today";

    /**
     * Bind View
     */
    private View view;
    /**
     * 标题集合
     */
    private final List<String> titleList;
    /**
     * 片段集合
     */
    private final List<Fragment> fragmentList;

    {
        titleList = new ArrayList<>();
        titleList.add("今天");
        titleList.add("本周");
        fragmentList = new ArrayList<>();
        fragmentList.add(new TodayFragment());
        fragmentList.add(new ThisWeekFragment());
    }

    static class ViewHolder {
        @BindView(R.id.tl)
        TabLayout tl;
        @BindView(R.id.vp)
        ViewPager vp;

        ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Log.d(TAG, "on Create View");
        ViewHolder holder;
        if (view != null) {
            holder = (ViewHolder) view.getTag();
        } else {
            view = inflater.inflate(R.layout.fragment_class_schedule, container, false);
            holder = new ViewHolder(view);
            view.setTag(holder);
        }
        ThemeChangeUtil.setTabLayoutColor(requireContext(), holder.tl);
        initData(holder);
        //设置默认展示页面
        if (!TODAY.equals(PreferenceManager.getDefaultSharedPreferences(requireContext()).getString(DEFAULT_SHOW_MAIN_FRAGMENT, TODAY))) {
            holder.vp.setCurrentItem(1);
            Objects.requireNonNull(holder.tl.getTabAt(1)).select();
        }
        return view;
    }

    private void initData(ViewHolder holder) {
        //预加载
        holder.vp.setOffscreenPageLimit(fragmentList.size());
        holder.vp.setAdapter(new FragmentStatePagerAdapter(getChildFragmentManager()) {

            @Override
            public int getCount() {
                return fragmentList.size();
            }

            @Override
            public Fragment getItem(int position) {
                return fragmentList.get(position);
            }

            @Nullable
            @Override
            public CharSequence getPageTitle(int position) {
                return titleList.get(position);
            }

            @Override
            public int getItemPosition(@NonNull Object object) {
                Log.d(TAG, "getItemPosition: " + object);
                return POSITION_NONE;
            }

            @Override
            public void restoreState(Parcelable state, ClassLoader loader) {
                try {
                    super.restoreState(state, loader);
                } catch (NullPointerException e) {
                    // null caught
                }
            }
        });
        holder.tl.setupWithViewPager(holder.vp);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        Log.d(TAG, "on Create");
        super.onCreate(savedInstanceState);
        EventBus.getDefault().register(this);
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "on Destroy");
        EventBus.getDefault().unregister(this);
        super.onDestroy();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(EventEntity eventEntity) {
        switch (eventEntity.getId()) {
            case APP_COLOR_CHANGE: {
                ThemeChangeUtil.setTabLayoutColor(requireContext(), ((ViewHolder) view.getTag()).tl);
                break;
            }
            case REFRESH_CLASS_SCHEDULE_FRAGMENT: {
                ViewHolder holder = (ViewHolder) view.getTag();
                PagerAdapter adapter = holder.vp.getAdapter();
                if (adapter == null) {
                    Toast.makeText(requireContext(), "未找到适配器，尝试重新打开APP解决此问题", Toast.LENGTH_LONG).show();
                    break;
                }
                adapter.notifyDataSetChanged();
                break;
            }
            default:
        }
    }
}

File: app/src/main/java/top/itning/yunshuclassschedule/ui/fragment/ClassScheduleFragment.kt
package top.itning.yunshuclassschedule.ui.fragment

import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.Nullable
import androidx.annotation.Px
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.preference.PreferenceManager
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import kotlinx.android.synthetic.main.fragment_class_schedule.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import top.itning.yunshuclassschedule.R
import top.itning.yunshuclassschedule.common.ConstantPool
import top.itning.yunshuclassschedule.entity.EventEntity
import top.itning.yunshuclassschedule.ui.activity.MainActivity
import top.itning.yunshuclassschedule.ui.fragment.setting.SettingsFragment.Companion.DEFAULT_SHOW_MAIN_FRAGMENT
import top.itning.yunshuclassschedule.util.ThemeChangeUtil
import java.util.*

/**
 * 课程表
 *
 * @author itning
 */
class ClassScheduleFragment : Fragment() {

    /**
     * 标题集合
     */
    private val titleList: MutableList<String>
    /**
     * 片段集合
     */
    private val fragmentList: MutableList<Fragment>

    init {
        titleList = ArrayList()
        titleList.add("今天")
        titleList.add("本周")
        fragmentList = ArrayList()
        fragmentList.add(TodayFragment())
        fragmentList.add(ThisWeekFragment())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Log.d(TAG, "on Create View")
        return inflater.inflate(R.layout.fragment_class_schedule, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        ThemeChangeUtil.setTabLayoutColor(requireContext(), tl)
        initData()
        //设置默认展示页面
        if (TODAY != PreferenceManager.getDefaultSharedPreferences(requireContext()).getString(DEFAULT_SHOW_MAIN_FRAGMENT, TODAY)) {
            vp.currentItem = 1
            tl.getTabAt(1)?.select()
        }
    }

    private fun initData() {
        //预加载
        vp.offscreenPageLimit = fragmentList.size
        vp.adapter = object : FragmentStatePagerAdapter(childFragmentManager) {

            override fun getCount(): Int {
                return fragmentList.size
            }

            override fun getItem(position: Int): Fragment {
                return fragmentList[position]
            }

            @Nullable
            override fun getPageTitle(position: Int): CharSequence {
                return titleList[position]
            }

            override fun getItemPosition(`object`: Any): Int {
                Log.d(TAG, "getItemPosition: $`object`")
                return PagerAdapter.POSITION_NONE
            }

            override fun restoreState(state: Parcelable?, loader: ClassLoader?) {
                try {
                    super.restoreState(state, loader)
                } catch (e: Exception) {
                    // null caught
                }

            }
        }
        vp.clearOnPageChangeListeners()
        vp.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, @Px positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                if (position == 0) {
                    EventBus.getDefault().post(EventEntity(ConstantPool.Int.TOOLBAR_TITLE_CHANGE, MainActivity.ACTION_BAR_TITLE_FORMAT.format(Date())))
                } else {
                    EventBus.getDefault().post(EventEntity(ConstantPool.Int.TOOLBAR_TITLE_CHANGE, "第${MainActivity.tempNumberOfWeek}周"))
                }
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })
        tl.setupWithViewPager(vp)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(TAG, "on Create")
        super.onCreate(savedInstanceState)
        EventBus.getDefault().register(this)
    }

    override fun onDestroy() {
        Log.d(TAG, "on Destroy")
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(eventEntity: EventEntity) {
        when (eventEntity.id) {
            ConstantPool.Int.APP_COLOR_CHANGE -> {
                ThemeChangeUtil.setTabLayoutColor(requireContext(), tl)
            }
            ConstantPool.Int.REFRESH_CLASS_SCHEDULE_FRAGMENT -> {
                val adapter = vp.adapter
                if (adapter == null) {
                    Toast.makeText(requireContext(), "未找到适配器，尝试重新打开APP解决此问题", Toast.LENGTH_LONG).show()
                    return
                }
                adapter.notifyDataSetChanged()
            }
            else -> {
            }
        }
    }

    companion object {
        private const val TAG = "ClassScheduleFragment"
        private const val TODAY = "today"
    }
}
